aptos-move/mvhashmap/src/versioned_data.rs

```diff
@@ -0,0 +1,294 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::types::{Incarnation, MVDataError, MVDataOutput, TxnIndex, Version};
+use aptos_aggregator::{
+    delta_change_set::{deserialize, DeltaOp},
+    transaction::AggregatorValue,
+};
+use aptos_infallible::Mutex;
+use aptos_types::write_set::TransactionWrite;
+use crossbeam::utils::CachePadded;
+use dashmap::DashMap;
+use std::{
+    collections::btree_map::BTreeMap,
+    hash::Hash,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc,
+    },
+};
+
+const FLAG_DONE: usize = 0;
+const FLAG_ESTIMATE: usize = 1;
+
+/// Every entry in shared multi-version data-structure has an "estimate" flag
+/// and some content.
+struct Entry<V> {
+    /// Used to mark the entry as a "write estimate". Even though the entry
+    /// lives inside the DashMap and the entry access will have barriers, we
+    /// still make the flag Atomic to provide acq/rel semantics on its own.
+    flag: AtomicUsize,
+
+    /// Actual contents.
+    pub cell: EntryCell<V>,
+}
+
+/// Represents the content of a single entry in multi-version data-structure.
+enum EntryCell<V> {
+    /// Recorded in the shared multi-version data-structure for each write. It
+    /// has: 1) Incarnation number of the transaction that wrote the entry (note
+    /// that TxnIndex is part of the key and not recorded here), 2) actual data
+    /// stored in a shared pointer (to ensure ownership and avoid clones).
+    Write(Incarnation, Arc<V>),
+
+    /// Recorded in the shared multi-version data-structure for each delta.
+    Delta(DeltaOp),
+}
+
+/// A VersionedValue internally contains a BTreeMap from indices of transactions
+/// that update the given access path alongside the corresponding entries.
+struct VersionedValue<V> {
+    versioned_map: BTreeMap<TxnIndex, CachePadded<Entry<V>>>,
+
+    // Note: this can cache base (storage) value in Option<u128> to facilitate
+    // aggregator validation & reading in the future, if needed.
+    contains_delta: bool,
```

Comment:
Unrelated to this diff, but remind me why we need this bool if we have this information in the delta_keys vector?

Reply:
it's currently just saving the unnecessary locking on delta_keys, and allows that to be a Vec. not too important, but also relevant meta-data that we can store in an entry so why not.